{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Explore h5ad files from TAHOE-100M using Scanpy\n",
    "#### Author: Leonardo Agasso, University of Turin (email: leonardo.agasso@unito.it)\n",
    "___________________________________________________________________________________________"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Import libraries"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pandas\n",
    "import numpy as np \n",
    "import matplotlib.pyplot as plt\n",
    "import scanpy as sc\n",
    "import os\n",
    "import sys"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Here you need to define the relative path to the file you want to open. In this case the file name indicates that the cells come from Plate 1, the cancer cell-line is \"HEC-1-A\" (human endometrial adenocarcinoma), only protein-coding genes were filtered and 15 random drugs are taken into account."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "n_plate = 1\n",
    "cell_line = 'HEC-1-A'\n",
    "biotype = 'protein_coding'\n",
    "\n",
    "h5ad_file_path = f'../../dataset/plate{str(n_plate)}_cell_line_{cell_line}.{biotype}.15_rnd_drugs.h5ad'"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Scanpy has a module (*sc.read_h5ad*) that allows to read h5ad files and store them in a specific object called *AnnData* (adata for brevity). Annotation data is the standard format used by Scanpy to store single-cell data."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(36320, 19915)"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata = sc.read_h5ad(h5ad_file_path)\n",
    "adata.shape"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Our file is quite large (470 Mb) but the annotation data handle large matrices very efficiently. The *shape* attribute of the *adata* object shows the number of cells and genes in the dataset, in this case we have 36,320 cells (*observables* in scanpy) and 19,915 genes (*variables* in scanpy)."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We can take a look at what these genes and cells are stored in the *adata* object. Originally the data from the Tahoe-100M project didn't have the biotype, I labeled the genes according to the latest release of [GENCODE](https://www.gencodegenes.org/human/) in order to filter only protein-coding genes. A gene is typically adressed by its Gene Name (that is standardized by the HUGO Gene Nomenclature Committee; examples are TSPAN6, HOXB3 etc.) and/or by its Ensembl ID (a univocal identifier like ENSG00000291316, ENSG00000291313 etc.)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>biotype</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gene_name</th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>TSPAN6</th>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TNMD</th>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>DPM1</th>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SCYL3</th>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>FGR</th>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000291313</th>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000291314</th>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000291315</th>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000291316</th>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TMEM276</th>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>19915 rows × 1 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                        biotype\n",
       "gene_name                      \n",
       "TSPAN6           protein_coding\n",
       "TNMD             protein_coding\n",
       "DPM1             protein_coding\n",
       "SCYL3            protein_coding\n",
       "FGR              protein_coding\n",
       "...                         ...\n",
       "ENSG00000291313  protein_coding\n",
       "ENSG00000291314  protein_coding\n",
       "ENSG00000291315  protein_coding\n",
       "ENSG00000291316  protein_coding\n",
       "TMEM276          protein_coding\n",
       "\n",
       "[19915 rows x 1 columns]"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata.var"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Cells have much more information in this dataset. The authors of the original work provide a table to understand the meaning of each column."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sample</th>\n",
       "      <th>gene_count</th>\n",
       "      <th>tscp_count</th>\n",
       "      <th>mread_count</th>\n",
       "      <th>drugname_drugconc</th>\n",
       "      <th>drug</th>\n",
       "      <th>cell_line</th>\n",
       "      <th>sublibrary</th>\n",
       "      <th>BARCODE</th>\n",
       "      <th>pcnt_mito</th>\n",
       "      <th>S_score</th>\n",
       "      <th>G2M_score</th>\n",
       "      <th>phase</th>\n",
       "      <th>pass_filter</th>\n",
       "      <th>cell_name</th>\n",
       "      <th>plate</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>BARCODE_SUB_LIB_ID</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>06_008_135-lib_841</th>\n",
       "      <td>smp_1500</td>\n",
       "      <td>1830</td>\n",
       "      <td>2929</td>\n",
       "      <td>3481</td>\n",
       "      <td>[('Capivasertib', 0.05, 'uM')]</td>\n",
       "      <td>Capivasertib</td>\n",
       "      <td>CVCL_0293</td>\n",
       "      <td>lib_841</td>\n",
       "      <td>06_008_135</td>\n",
       "      <td>0.136224</td>\n",
       "      <td>-0.119048</td>\n",
       "      <td>-0.094322</td>\n",
       "      <td>G1</td>\n",
       "      <td>full</td>\n",
       "      <td>HEC-1-A</td>\n",
       "      <td>plate1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>06_009_187-lib_841</th>\n",
       "      <td>smp_1500</td>\n",
       "      <td>995</td>\n",
       "      <td>1330</td>\n",
       "      <td>1573</td>\n",
       "      <td>[('Capivasertib', 0.05, 'uM')]</td>\n",
       "      <td>Capivasertib</td>\n",
       "      <td>CVCL_0293</td>\n",
       "      <td>lib_841</td>\n",
       "      <td>06_009_187</td>\n",
       "      <td>0.083459</td>\n",
       "      <td>0.071429</td>\n",
       "      <td>-0.080586</td>\n",
       "      <td>S</td>\n",
       "      <td>full</td>\n",
       "      <td>HEC-1-A</td>\n",
       "      <td>plate1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>06_015_001-lib_841</th>\n",
       "      <td>smp_1500</td>\n",
       "      <td>3320</td>\n",
       "      <td>6043</td>\n",
       "      <td>7075</td>\n",
       "      <td>[('Capivasertib', 0.05, 'uM')]</td>\n",
       "      <td>Capivasertib</td>\n",
       "      <td>CVCL_0293</td>\n",
       "      <td>lib_841</td>\n",
       "      <td>06_015_001</td>\n",
       "      <td>0.088201</td>\n",
       "      <td>-0.061905</td>\n",
       "      <td>-0.136081</td>\n",
       "      <td>G1</td>\n",
       "      <td>full</td>\n",
       "      <td>HEC-1-A</td>\n",
       "      <td>plate1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>06_031_041-lib_841</th>\n",
       "      <td>smp_1500</td>\n",
       "      <td>1637</td>\n",
       "      <td>2373</td>\n",
       "      <td>2805</td>\n",
       "      <td>[('Capivasertib', 0.05, 'uM')]</td>\n",
       "      <td>Capivasertib</td>\n",
       "      <td>CVCL_0293</td>\n",
       "      <td>lib_841</td>\n",
       "      <td>06_031_041</td>\n",
       "      <td>0.093552</td>\n",
       "      <td>0.019048</td>\n",
       "      <td>0.068498</td>\n",
       "      <td>G2M</td>\n",
       "      <td>full</td>\n",
       "      <td>HEC-1-A</td>\n",
       "      <td>plate1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>06_031_049-lib_841</th>\n",
       "      <td>smp_1500</td>\n",
       "      <td>679</td>\n",
       "      <td>852</td>\n",
       "      <td>1029</td>\n",
       "      <td>[('Capivasertib', 0.05, 'uM')]</td>\n",
       "      <td>Capivasertib</td>\n",
       "      <td>CVCL_0293</td>\n",
       "      <td>lib_841</td>\n",
       "      <td>06_031_049</td>\n",
       "      <td>0.103286</td>\n",
       "      <td>0.023810</td>\n",
       "      <td>-0.009158</td>\n",
       "      <td>S</td>\n",
       "      <td>full</td>\n",
       "      <td>HEC-1-A</td>\n",
       "      <td>plate1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>93_183_069-lib_912</th>\n",
       "      <td>smp_1587</td>\n",
       "      <td>2377</td>\n",
       "      <td>3829</td>\n",
       "      <td>4561</td>\n",
       "      <td>[('Anastrozole', 0.05, 'uM')]</td>\n",
       "      <td>Anastrozole</td>\n",
       "      <td>CVCL_0293</td>\n",
       "      <td>lib_912</td>\n",
       "      <td>93_183_069</td>\n",
       "      <td>0.068686</td>\n",
       "      <td>0.251196</td>\n",
       "      <td>-0.075727</td>\n",
       "      <td>S</td>\n",
       "      <td>full</td>\n",
       "      <td>HEC-1-A</td>\n",
       "      <td>plate1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>93_185_090-lib_912</th>\n",
       "      <td>smp_1587</td>\n",
       "      <td>846</td>\n",
       "      <td>1100</td>\n",
       "      <td>1277</td>\n",
       "      <td>[('Anastrozole', 0.05, 'uM')]</td>\n",
       "      <td>Anastrozole</td>\n",
       "      <td>CVCL_0293</td>\n",
       "      <td>lib_912</td>\n",
       "      <td>93_185_090</td>\n",
       "      <td>0.087273</td>\n",
       "      <td>-0.005013</td>\n",
       "      <td>0.024291</td>\n",
       "      <td>G2M</td>\n",
       "      <td>full</td>\n",
       "      <td>HEC-1-A</td>\n",
       "      <td>plate1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>93_187_075-lib_912</th>\n",
       "      <td>smp_1587</td>\n",
       "      <td>958</td>\n",
       "      <td>1308</td>\n",
       "      <td>1563</td>\n",
       "      <td>[('Anastrozole', 0.05, 'uM')]</td>\n",
       "      <td>Anastrozole</td>\n",
       "      <td>CVCL_0293</td>\n",
       "      <td>lib_912</td>\n",
       "      <td>93_187_075</td>\n",
       "      <td>0.074924</td>\n",
       "      <td>-0.048075</td>\n",
       "      <td>-0.038001</td>\n",
       "      <td>G1</td>\n",
       "      <td>full</td>\n",
       "      <td>HEC-1-A</td>\n",
       "      <td>plate1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>93_187_185-lib_912</th>\n",
       "      <td>smp_1587</td>\n",
       "      <td>1503</td>\n",
       "      <td>2138</td>\n",
       "      <td>2532</td>\n",
       "      <td>[('Anastrozole', 0.05, 'uM')]</td>\n",
       "      <td>Anastrozole</td>\n",
       "      <td>CVCL_0293</td>\n",
       "      <td>lib_912</td>\n",
       "      <td>93_187_185</td>\n",
       "      <td>0.068288</td>\n",
       "      <td>0.027683</td>\n",
       "      <td>0.048951</td>\n",
       "      <td>G2M</td>\n",
       "      <td>full</td>\n",
       "      <td>HEC-1-A</td>\n",
       "      <td>plate1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>93_188_148-lib_912</th>\n",
       "      <td>smp_1587</td>\n",
       "      <td>1391</td>\n",
       "      <td>1952</td>\n",
       "      <td>2319</td>\n",
       "      <td>[('Anastrozole', 0.05, 'uM')]</td>\n",
       "      <td>Anastrozole</td>\n",
       "      <td>CVCL_0293</td>\n",
       "      <td>lib_912</td>\n",
       "      <td>93_188_148</td>\n",
       "      <td>0.072746</td>\n",
       "      <td>-0.081795</td>\n",
       "      <td>-0.061649</td>\n",
       "      <td>G1</td>\n",
       "      <td>full</td>\n",
       "      <td>HEC-1-A</td>\n",
       "      <td>plate1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>36320 rows × 16 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                      sample  gene_count  tscp_count  mread_count  \\\n",
       "BARCODE_SUB_LIB_ID                                                  \n",
       "06_008_135-lib_841  smp_1500        1830        2929         3481   \n",
       "06_009_187-lib_841  smp_1500         995        1330         1573   \n",
       "06_015_001-lib_841  smp_1500        3320        6043         7075   \n",
       "06_031_041-lib_841  smp_1500        1637        2373         2805   \n",
       "06_031_049-lib_841  smp_1500         679         852         1029   \n",
       "...                      ...         ...         ...          ...   \n",
       "93_183_069-lib_912  smp_1587        2377        3829         4561   \n",
       "93_185_090-lib_912  smp_1587         846        1100         1277   \n",
       "93_187_075-lib_912  smp_1587         958        1308         1563   \n",
       "93_187_185-lib_912  smp_1587        1503        2138         2532   \n",
       "93_188_148-lib_912  smp_1587        1391        1952         2319   \n",
       "\n",
       "                                 drugname_drugconc          drug  cell_line  \\\n",
       "BARCODE_SUB_LIB_ID                                                            \n",
       "06_008_135-lib_841  [('Capivasertib', 0.05, 'uM')]  Capivasertib  CVCL_0293   \n",
       "06_009_187-lib_841  [('Capivasertib', 0.05, 'uM')]  Capivasertib  CVCL_0293   \n",
       "06_015_001-lib_841  [('Capivasertib', 0.05, 'uM')]  Capivasertib  CVCL_0293   \n",
       "06_031_041-lib_841  [('Capivasertib', 0.05, 'uM')]  Capivasertib  CVCL_0293   \n",
       "06_031_049-lib_841  [('Capivasertib', 0.05, 'uM')]  Capivasertib  CVCL_0293   \n",
       "...                                            ...           ...        ...   \n",
       "93_183_069-lib_912   [('Anastrozole', 0.05, 'uM')]   Anastrozole  CVCL_0293   \n",
       "93_185_090-lib_912   [('Anastrozole', 0.05, 'uM')]   Anastrozole  CVCL_0293   \n",
       "93_187_075-lib_912   [('Anastrozole', 0.05, 'uM')]   Anastrozole  CVCL_0293   \n",
       "93_187_185-lib_912   [('Anastrozole', 0.05, 'uM')]   Anastrozole  CVCL_0293   \n",
       "93_188_148-lib_912   [('Anastrozole', 0.05, 'uM')]   Anastrozole  CVCL_0293   \n",
       "\n",
       "                   sublibrary     BARCODE  pcnt_mito   S_score  G2M_score  \\\n",
       "BARCODE_SUB_LIB_ID                                                          \n",
       "06_008_135-lib_841    lib_841  06_008_135   0.136224 -0.119048  -0.094322   \n",
       "06_009_187-lib_841    lib_841  06_009_187   0.083459  0.071429  -0.080586   \n",
       "06_015_001-lib_841    lib_841  06_015_001   0.088201 -0.061905  -0.136081   \n",
       "06_031_041-lib_841    lib_841  06_031_041   0.093552  0.019048   0.068498   \n",
       "06_031_049-lib_841    lib_841  06_031_049   0.103286  0.023810  -0.009158   \n",
       "...                       ...         ...        ...       ...        ...   \n",
       "93_183_069-lib_912    lib_912  93_183_069   0.068686  0.251196  -0.075727   \n",
       "93_185_090-lib_912    lib_912  93_185_090   0.087273 -0.005013   0.024291   \n",
       "93_187_075-lib_912    lib_912  93_187_075   0.074924 -0.048075  -0.038001   \n",
       "93_187_185-lib_912    lib_912  93_187_185   0.068288  0.027683   0.048951   \n",
       "93_188_148-lib_912    lib_912  93_188_148   0.072746 -0.081795  -0.061649   \n",
       "\n",
       "                   phase pass_filter cell_name   plate  \n",
       "BARCODE_SUB_LIB_ID                                      \n",
       "06_008_135-lib_841    G1        full   HEC-1-A  plate1  \n",
       "06_009_187-lib_841     S        full   HEC-1-A  plate1  \n",
       "06_015_001-lib_841    G1        full   HEC-1-A  plate1  \n",
       "06_031_041-lib_841   G2M        full   HEC-1-A  plate1  \n",
       "06_031_049-lib_841     S        full   HEC-1-A  plate1  \n",
       "...                  ...         ...       ...     ...  \n",
       "93_183_069-lib_912     S        full   HEC-1-A  plate1  \n",
       "93_185_090-lib_912   G2M        full   HEC-1-A  plate1  \n",
       "93_187_075-lib_912    G1        full   HEC-1-A  plate1  \n",
       "93_187_185-lib_912   G2M        full   HEC-1-A  plate1  \n",
       "93_188_148-lib_912    G1        full   HEC-1-A  plate1  \n",
       "\n",
       "[36320 rows x 16 columns]"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata.obs"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Here's the table formatted with consistent spacing, copied from the [official repository](https://github.com/ArcInstitute/arc-virtual-cell-atlas/blob/main/tahoe-100/README.md?plain=1) of the project.\n",
    "\n",
    "| Column Name            | Description                                                            |\n",
    "|------------------------|------------------------------------------------------------------------|\n",
    "| **plate**              | Plate identifier                                                       |\n",
    "| **BARCODE_SUB_LIB_ID** | Cell identifier                                                        |\n",
    "| **sample**             | Unique treatment identifier, distinguishes replicated treatments       |\n",
    "| **gene_count**         | Number of genes with at least one count                                |\n",
    "| **tscp_count**         | Number of transcripts, aka UMI count                                   |\n",
    "| **mread_count**        | Number of reads per cell                                               |\n",
    "| **drugname_drugconc**  | Drug name, concentration, and concentration unit                       |\n",
    "| **drug**               | Drug name, parsed out from the `drugname_drugconc` field               |\n",
    "| **cell_line**          | Cell line Cellosaurus identifier                                       |\n",
    "| **sublibrary**         | Sublibrary ID (related to library prep and sequencing)                 |\n",
    "| **BARCODE**            | Barcode ID                                                             |\n",
    "| **pcnt_mito**          | Percentage of mitochondrial reads                                      |\n",
    "| **S_score**            | Inferred S phase score                                                 |\n",
    "| **G2M_score**          | Inferred G2M score                                                     |\n",
    "| **phase**              | Inferred cell cycle phase                                              |\n",
    "| **pass_filter**        | \"Full\" filters are more stringent on `gene_count` and `tscp_count`     |\n",
    "| **cell_name**          | Commonly-used cell name (related to the `cell_line` field)             |\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "OhnomirnaEnv",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.14"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
